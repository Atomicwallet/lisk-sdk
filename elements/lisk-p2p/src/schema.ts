/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { Schema } from '@liskhq/lisk-codec';

export const nodeInfoSchema = {
	$id: '/nodeInfo',
	type: 'object',
	properties: {
		networkId: {
			dataType: 'string',
			fieldNumber: 1,
		},
		networkVersion: {
			dataType: 'string',
			fieldNumber: 2,
		},
		port: {
			dataType: 'uint32',
			fieldNumber: 3,
		},
		nonce: {
			dataType: 'string',
			fieldNumber: 4,
		},
		advertiseAddress: {
			dataType: 'boolean',
			fieldNumber: 5,
		},
		os: {
			dataType: 'string',
			fieldNumber: 6,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 7,
		},
	},
	required: ['networkId', 'networkVersion', 'port', 'nonce'],
};

export const peerInfoSchema = {
	$id: '/peerInfo',
	type: 'object',
	properties: {
		ipAddress: {
			dataType: 'string',
			fieldNumber: 1,
		},
		port: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		networkId: {
			dataType: 'string',
			fieldNumber: 3,
		},
		networkVersion: {
			dataType: 'string',
			fieldNumber: 4,
		},
		nonce: {
			dataType: 'string',
			fieldNumber: 5,
		},
		os: {
			dataType: 'string',
			fieldNumber: 6,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 7,
		},
	},
	required: ['ipAddress', 'port'],
};

export const defaultRPCSchemas = {
	peerInfo: peerInfoSchema,
	nodeInfo: nodeInfoSchema,
};

export const mergeCustomSchema = (
	baseSchema: Schema,
	customSchema: Schema,
): Schema => ({
	$id: `${baseSchema.$id}/custom`,
	type: 'object',
	properties: {
		...baseSchema.properties,
		...customSchema.properties,
	},
});
