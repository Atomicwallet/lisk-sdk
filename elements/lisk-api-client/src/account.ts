/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { codec } from '@liskhq/lisk-codec';
import { Channel, RegisteredSchemas } from './types';

export class Account {
	private readonly _channel: Channel;
	private readonly _schemas: RegisteredSchemas;

	public constructor(channel: Channel, schemas: RegisteredSchemas) {
		this._channel = channel;
		this._schemas = schemas;
	}

	public async get(address: Buffer): Promise<Record<string, unknown>> {
		const accountHex: string = await this._channel.invoke('app:getAccount', {
			address: address.toString('hex'),
		});

		return this.decode(Buffer.from(accountHex, 'hex'));
	}

	public encode(input: Record<string, unknown>): Buffer {
		return codec.encode(this._schemas.account, input);
	}

	public decode(input: Buffer): Record<string, unknown> {
		return codec.decode(this._schemas.account, input);
	}

	public toJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.toJSON(this._schemas.account, account);
	}

	public fromJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.fromJSON(this._schemas.account, account);
	}
}
