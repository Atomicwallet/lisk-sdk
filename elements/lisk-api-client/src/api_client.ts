/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { EventCallback, Channel, RegisteredSchemas, NodeInfo } from './types';
import { Node } from './node';

export class APIClient {
	private readonly _channel: Channel;
	private _schemas!: RegisteredSchemas;
	private _nodeInfo!: NodeInfo;
	private _node!: Node;

	public constructor(channel: Channel) {
		this._channel = channel;
	}

	public async init(): Promise<void> {
		this._node = new Node(this._channel);
		this._schemas = await this._channel.invoke<RegisteredSchemas>('app:getSchema');
		this._nodeInfo = await this._node.getNodeInfo();
	}

	public async disconnect(): Promise<void> {
		return this._channel.disconnect();
	}

	public async invoke<T = Record<string, unknown>>(
		actionName: string,
		params?: Record<string, unknown>,
	): Promise<T> {
		return this._channel.invoke(actionName, params);
	}

	public subscribe<T = Record<string, unknown>>(eventName: string, cb: EventCallback<T>): void {
		this._channel.subscribe(eventName, cb);
	}

	public get node(): Node {
		// FIXME: remove unnecessary code
		// eslint-disable-next-line no-console
		console.log(this._schemas, this._nodeInfo);
		return this._node;
	}
}
