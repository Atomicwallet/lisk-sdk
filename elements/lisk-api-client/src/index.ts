/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
export { createWSClient } from './ws_client';
export { createClient } from './client';
export { createAPIClient } from './create_api_client';
export { IPCChannel } from './ipc_channel';
