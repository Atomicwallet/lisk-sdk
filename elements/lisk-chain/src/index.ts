/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Chain, EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK } from './chain';
const events = { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK };

export { Account } from './account';
export * from './block_reward';
export { Chain, events };
export { baseBlockSchema } from './schema';
export { StateStore } from './state_store';
export { Slots } from './slots';
