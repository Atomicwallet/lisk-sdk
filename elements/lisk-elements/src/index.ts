/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
export * as apiClient from '../../lisk-api-client';
export * as cryptography from '../../lisk-cryptography';
export * as p2p from '../../lisk-p2p';
export * as passphrase from '../../lisk-passphrase';
export * as transactionPool from '../../lisk-transaction-pool';
export * as transactions from '../../lisk-transactions';
export * as utils from '../../lisk-utils';
export * as tree from '../../lisk-tree';
export * as validator from '../../lisk-validator';
export * as codec from '../../lisk-codec';
export * as db from '../../lisk-db';
export * as chain from '../../lisk-chain';
export * as bft from '../../lisk-bft';
export * as genesis from '../../lisk-genesis';
