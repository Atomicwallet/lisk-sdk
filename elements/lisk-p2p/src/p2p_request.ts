/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { RPCResponseAlreadySentError } from './errors';
import { P2PResponsePacketBufferData } from './types';

export interface RequestOptions {
	readonly procedure: string;
	readonly data: unknown;
	readonly id: string;
	readonly rate: number;
	productivity: {
		requestCounter: number;
		responseCounter: number;
		responseRate: number;
		lastResponded: number;
	};
}

export class P2PRequest {
	private readonly _procedure: string;
	private readonly _data: unknown;
	private readonly _respondCallback: (
		responseError?: Error,
		responseData?: P2PResponsePacketBufferData,
	) => void;
	private readonly _peerId: string;
	private _wasResponseSent: boolean;
	private readonly _rate: number;

	public constructor(
		options: RequestOptions,
		respondCallback: (responseError?: Error, responseData?: unknown) => void,
	) {
		this._procedure = options.procedure;
		this._data = options.data;
		this._peerId = options.id;
		this._rate = options.rate;
		// eslint-disable-next-line no-param-reassign
		options.productivity.requestCounter += 1;
		this._respondCallback = (
			responseError?: Error,
			responsePacket?: P2PResponsePacketBufferData,
		): void => {
			if (this._wasResponseSent) {
				throw new RPCResponseAlreadySentError(
					`A response has already been sent for the request procedure <<${options.procedure}>>`,
				);
			}
			this._wasResponseSent = true;
			// We assume peer performed useful work and update peer response rate
			if (!responseError && responsePacket) {
				// eslint-disable-next-line no-param-reassign
				options.productivity.lastResponded = Date.now();
				// eslint-disable-next-line no-param-reassign
				options.productivity.responseCounter += 1;
			}
			// eslint-disable-next-line no-param-reassign
			options.productivity.responseRate =
				options.productivity.responseCounter / options.productivity.requestCounter;

			let responsePacketBufferData: Buffer | undefined;
			if (responsePacket?.data && typeof responsePacket?.data === 'string') {
				responsePacketBufferData = Buffer.from(responsePacket.data, 'binary');
			}
			respondCallback(responseError, responsePacketBufferData);
		};
		this._wasResponseSent = false;
	}

	public get procedure(): string {
		return this._procedure;
	}

	public get data(): unknown {
		return this._data;
	}

	public get rate(): number {
		return this._rate;
	}

	public get peerId(): string {
		return this._peerId;
	}

	public get wasResponseSent(): boolean {
		return this._wasResponseSent;
	}

	public end(responseData: Buffer): void {
		const responsePacket: P2PResponsePacketBufferData = {
			data: responseData,
			peerId: this.peerId,
		};
		this._respondCallback(undefined, responsePacket);
	}

	public error(responseError: Error): void {
		this._respondCallback(responseError);
	}
}
