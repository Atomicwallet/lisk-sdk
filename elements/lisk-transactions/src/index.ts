/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { DelegateTransaction } from './10_delegate_transaction';
import { MultisignatureTransaction } from './12_multisignature_transaction';
import { VoteTransaction } from './13_vote_transaction';
import { UnlockTransaction } from './14_unlock_transaction';
import { ProofOfMisbehaviorTransaction } from './15_proof_of_misbehavior_transaction';
import { TransferTransaction } from './8_transfer_transaction';
import { BaseTransaction, StateStore } from './base_transaction';
import * as constants from './constants';
import { convertToAssetError, convertToTransactionError, TransactionError } from './errors';
import { createResponse, Status, TransactionResponse } from './response';
import { transactionInterface } from './schema';
import {
	getSigningBytes,
	validateTransactionSchema,
	signTransaction,
	signMultiSignatureTransaction,
} from './transaction_helper';
import { Account, TransactionJSON } from './types';
import {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSenderIdAndPublicKey,
	validateSignature,
	verifyMinRemainingBalance,
	verifyMultiSignatureTransaction,
} from './utils';

const exposedUtils = {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSignature,
	verifyMinRemainingBalance,
	validateSenderIdAndPublicKey,
	verifyMultiSignatureTransaction,
};

export {
	Account,
	BaseTransaction,
	getSigningBytes,
	StateStore,
	TransferTransaction,
	DelegateTransaction,
	VoteTransaction,
	MultisignatureTransaction,
	UnlockTransaction,
	createResponse,
	ProofOfMisbehaviorTransaction,
	signTransaction,
	signMultiSignatureTransaction,
	Status,
	TransactionResponse,
	TransactionJSON,
	TransactionError,
	validateTransactionSchema,
	transactionInterface,
	convertToAssetError,
	convertToTransactionError,
	constants,
	exposedUtils as utils,
};
