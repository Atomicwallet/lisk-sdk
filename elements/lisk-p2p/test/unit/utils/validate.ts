/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import {
	validatePeerAddress,
	validatePeerInfo,
	validateRPCRequest,
	validateProtocolMessage,
	validateNodeInfo,
	sanitizeIncomingPeerInfo,
	validatePeerInfoList,
} from '../../../src/utils';
import {
	ProtocolPeerInfo,
	P2PRequestPacket,
	P2PMessagePacket,
	P2PNodeInfo,
} from '../../../src/types';
import { constants } from '../../../src';

const {
	DEFAULT_MAX_PEER_DISCOVERY_RESPONSE_LENGTH,
	DEFAULT_MAX_PEER_INFO_SIZE,
	PEER_INFO_LIST_TOO_LONG_REASON,
	INVALID_PEER_INFO_LIST_REASON,
} = constants;

describe('utils/validate', () => {
	describe('#validatePeerInfo', () => {
		describe('for valid peer response object', () => {
			const peer: ProtocolPeerInfo = {
				ipAddress: '12.23.54.3',
				port: 5393,
			};

			const peerWithInvalidHeightValue: unknown = {
				ipAddress: '12.23.54.3',
				port: 5393,
			};

			it('should return P2PPeerInfo object', () => {
				expect(validatePeerInfo(sanitizeIncomingPeerInfo(peer), 10000)).toEqual(
					{
						peerId: '12.23.54.3:5393',
						ipAddress: '12.23.54.3',
						port: 5393,
						sharedState: {},
					},
				);
			});

			it('should return P2PPeerInfo object with height value set to 0', () => {
				expect(
					validatePeerInfo(
						sanitizeIncomingPeerInfo(peerWithInvalidHeightValue),
						10000,
					),
				).toEqual({
					peerId: '12.23.54.3:5393',
					ipAddress: '12.23.54.3',
					port: 5393,
					sharedState: {},
				});
			});
		});

		describe('for invalid peer response object', () => {
			it('should throw an InvalidPeer error for invalid peer', () => {
				const peerInvalid: unknown = null;

				expect(
					validatePeerInfo.bind(
						null,
						sanitizeIncomingPeerInfo(peerInvalid),
						10000,
					),
				).toThrow('Invalid peer object');
			});

			it('should throw if PeerInfo is too big', () => {
				const maximumPeerInfoSizeInBytes = 10;
				const peer: ProtocolPeerInfo = {
					ipAddress: '12.23.54.3',
					port: 5393,
				};

				expect(
					validatePeerInfo.bind(
						null,
						sanitizeIncomingPeerInfo(peer),
						maximumPeerInfoSizeInBytes,
					),
				).toThrow(
					`PeerInfo is larger than the maximum allowed size ${maximumPeerInfoSizeInBytes} bytes`,
				);
			});

			it('should throw InvalidPeer error for invalid peer ipAddress or port', () => {
				const peerInvalid: unknown = {
					port: 53937888,
					height: '23232',
					discoveredInfo: {
						os: 'darwin',
					},
				};

				expect(
					validatePeerInfo.bind(
						null,
						sanitizeIncomingPeerInfo(peerInvalid),
						10000,
					),
				).toThrow('Invalid peer ipAddress or port');
			});
		});
	});

	describe('#validatePeerInfoList', () => {
		let generatePeerInfoResponse: any = {
			peers: [],
		};
		beforeEach(() => {
			generatePeerInfoResponse = {
				peers: [],
			};

			generatePeerInfoResponse.peers = [...Array(3)].map(() => ({
				ipAddress: '128.127.126.125',
				port: 5000,
			}));
		});

		describe('when PeerInfo list is valid', () => {
			it('should return P2PPeerInfo array', () => {
				expect(
					validatePeerInfoList(
						generatePeerInfoResponse,
						DEFAULT_MAX_PEER_DISCOVERY_RESPONSE_LENGTH,
						DEFAULT_MAX_PEER_INFO_SIZE,
					),
				).toHaveLength(generatePeerInfoResponse.peers.length);
			});
		});

		describe('when rawBasicPeerInfoList list is falsy', () => {
			it('should throw an Error', () => {
				generatePeerInfoResponse = undefined;

				expect(
					validatePeerInfoList.bind(
						null,
						generatePeerInfoResponse,
						DEFAULT_MAX_PEER_DISCOVERY_RESPONSE_LENGTH,
						DEFAULT_MAX_PEER_INFO_SIZE,
					),
				).toThrow(INVALID_PEER_INFO_LIST_REASON);
			});
		});

		describe('when PeerInfo list is not an array', () => {
			it('should throw an Error', () => {
				generatePeerInfoResponse.peers = 'fizzBuzz';

				expect(
					validatePeerInfoList.bind(
						null,
						generatePeerInfoResponse,
						DEFAULT_MAX_PEER_DISCOVERY_RESPONSE_LENGTH,
						DEFAULT_MAX_PEER_INFO_SIZE,
					),
				).toThrow(INVALID_PEER_INFO_LIST_REASON);
			});
		});

		describe('when PeerInfo list os too long', () => {
			it('should throw an Error', () => {
				expect(
					validatePeerInfoList.bind(
						null,
						generatePeerInfoResponse,
						generatePeerInfoResponse.peers.length - 1,
						DEFAULT_MAX_PEER_INFO_SIZE,
					),
				).toThrow(PEER_INFO_LIST_TOO_LONG_REASON);
			});
		});

		describe('when PeerInfo list has falsy PeerInfo', () => {
			it('should return P2PPeerInfo array', () => {
				generatePeerInfoResponse.peers.push(undefined);

				expect(
					validatePeerInfoList.bind(
						null,
						generatePeerInfoResponse,
						DEFAULT_MAX_PEER_DISCOVERY_RESPONSE_LENGTH,
						DEFAULT_MAX_PEER_INFO_SIZE,
					),
				).toThrow('Invalid peer object');
			});
		});
	});

	describe('#validateNodeInfo', () => {
		describe('when NodeInfo is larger than maximum allowed size', () => {
			const maximumSize = 10;

			const NodeInfo: P2PNodeInfo = {
				os: '12.23.54.3',
				networkId: '12.23.54.3',
				port: 5393,
				version: '1.1.2',
				protocolVersion: '1.1',
				options: {
					foo: 'bar',
					fizz: 'buzz',
				},
				nonce: 'nonce678',
				advertiseAddress: true,
			};

			it('should throw Invalid NodeInfo maximum allowed size error', () => {
				expect(validateNodeInfo.bind(null, NodeInfo, maximumSize)).toThrow(
					`Invalid NodeInfo was larger than the maximum allowed ${maximumSize} bytes`,
				);
			});
		});
	});

	describe('#validatePeerAddress', () => {
		it('should return true for correct IPv4', () => {
			const peer = {
				ipAddress: '12.12.12.12',
				port: 4001,
			};

			expect(validatePeerAddress(peer.ipAddress, peer.port)).toBe(true);
		});

		it('should return true for correct IPv6', () => {
			const peer = {
				ipAddress: '2001:0db8:85a3:0000:0000:8a2e:0370:7334',
				port: 4001,
			};

			expect(validatePeerAddress(peer.ipAddress, peer.port)).toBe(true);
		});

		it('should return false for incorrect ipAddress', () => {
			const peerWithIncorrectIp = {
				ipAddress: '12.12.hh12.12',
				port: 4001,
			};

			expect(
				validatePeerAddress(
					peerWithIncorrectIp.ipAddress,
					peerWithIncorrectIp.port,
				),
			).toBe(false);
		});

		it('should return false for incorrect port', () => {
			const peerWithIncorrectPort = {
				ipAddress: '12.12.12.12',
				port: NaN,
			};

			expect(
				validatePeerAddress(
					peerWithIncorrectPort.ipAddress,
					peerWithIncorrectPort.port,
				),
			).toBe(false);
		});
	});

	describe('#validateRPCRequest', () => {
		const validRPCRequest: unknown = {
			data: {},
			procedure: 'list',
		};
		let validatedRPCRequest: P2PRequestPacket;

		beforeEach(() => {
			validatedRPCRequest = validateRPCRequest(validRPCRequest);
		});

		it('should throw an error for an invalid procedure value', () => {
			expect(validateRPCRequest.bind(validateRPCRequest, undefined)).toThrow(
				'Invalid request',
			);
		});

		it('should throw an error for an invalid procedure value with object', () => {
			const inValidRequest: unknown = {
				data: {},
				procedure: {},
			};

			expect(
				validateRPCRequest.bind(validateRPCRequest, inValidRequest),
			).toThrow('Request procedure name is not a string');
		});

		it('should pass and return an object', () => {
			expect(validatedRPCRequest).toEqual(expect.any(Object));
		});

		it('should return a valid rpc request', () => {
			expect(validatedRPCRequest).toMatchObject({
				procedure: expect.any(String),
				data: expect.any(Object),
			});
		});
	});

	describe('#validateProtocolMessage', () => {
		const validProtocolMessage: unknown = {
			data: {},
			event: 'newPeer',
		};
		let returnedValidatedMessage: P2PMessagePacket;

		beforeEach(() => {
			returnedValidatedMessage = validateProtocolMessage(validProtocolMessage);
		});

		it('should throw an error for an invalid event value type', () => {
			expect(
				validateProtocolMessage.bind(validateProtocolMessage, undefined),
			).toThrow('Invalid message');
		});

		it('should throw an error for an invalid event value type with number', () => {
			const inValidMessage: unknown = {
				data: {},
				event: 6788,
			};
			expect(
				validateProtocolMessage.bind(validateProtocolMessage, inValidMessage),
			).toThrow('Protocol message is not a string');
		});

		it('should return an object', () => {
			expect(returnedValidatedMessage).toEqual(expect.any(Object));
		});

		it('should return a valid protocol message object', () => {
			expect(returnedValidatedMessage).toMatchObject({
				data: expect.any(Object),
				event: 'newPeer',
			});
		});
	});
});
