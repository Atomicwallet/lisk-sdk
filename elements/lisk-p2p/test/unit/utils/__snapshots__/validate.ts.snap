// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/validate #validateProtocolMessage should return a valid protocol message object 1`] = `
Object {
  "data": Any<Object>,
  "event": "newPeer",
}
`;

exports[`utils/validate #validateRPCRequest should return a valid rpc request 1`] = `
Object {
  "data": Any<Object>,
  "procedure": Any<String>,
  "type": Any<String>,
}
`;
